client/X11/xf_cliprdr.c

```diff
@@ -1434,9 +1575,133 @@ xf_cliprdr_server_format_data_response(CliprdrClientContext* context,
+					xfCliprdrFuseInode* rootNode =
+					    (xfCliprdrFuseInode*)calloc(1, sizeof(xfCliprdrFuseInode));
+
+					rootNode->ino = 1;
+					rootNode->parent_ino = 1;
+					rootNode->name = _strdup("/");
+					rootNode->st_mode = S_IFDIR | 0755;
+					rootNode->child_inos = ArrayList_New(TRUE);
+					rootNode->st_mtim.tv_sec = time(NULL);
+					ArrayList_Add(clipboard->ino_list, rootNode);
+
+					mapDir = HashTable_New(TRUE);
+					mapDir->keyFree = HashTable_StringFree;
+					mapDir->keyClone = HashTable_StringClone;
+					mapDir->keyCompare = HashTable_StringCompare;
+					mapDir->hash = HashTable_StringHash;
+
+					/* here we assume that parent folder always appears before its children */
+					for (lindex = 0; lindex < count; lindex++)
+					{
+						xfCliprdrFuseInode* inode =
+						    (xfCliprdrFuseInode*)calloc(1, sizeof(xfCliprdrFuseInode));
+						size_t curLen = _wcsnlen(descriptors[lindex].cFileName,
+						                         ARRAYSIZE(descriptors[lindex].cFileName));
+						char* curName = NULL;
+						int newLen = ConvertFromUnicode(CP_UTF8, 0, descriptors[lindex].cFileName,
+						                                (int)curLen, &curName, 0, NULL, NULL);
+						char* split_point = xf_cliprdr_fuse_split_basename(curName, newLen);
+						char* dirName = NULL;
+						char* baseName = NULL;
+						UINT64 ticks;
+						xfCliprdrFuseInode* parent;
+
+						inode->lindex = lindex;
+						inode->ino = lindex + 2;
+
+						if (split_point == NULL)
+						{
+							baseName = _strdup(curName);
+							inode->parent_ino = 1;
+							inode->name = baseName;
+							ArrayList_Add(rootNode->child_inos, (void*)inode->ino);
+						}
+						else
+						{
+							dirName = calloc(split_point - curName + 1, sizeof(char));
+							_snprintf(dirName, split_point - curName + 1, "%s", curName);
+							/* drop last '\\' */
+							baseName = _strdup(split_point + 1);
+
+							parent = (xfCliprdrFuseInode*)HashTable_GetItemValue(mapDir, dirName);
+							inode->parent_ino = parent->ino;
+							inode->name = baseName;
+							ArrayList_Add(parent->child_inos, (void*)inode->ino);
+							free(dirName);
+						}
+
+						if ((descriptors[lindex].dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+						{
+							inode->st_mode = S_IFDIR | 0755;
+							inode->child_inos = ArrayList_New(TRUE);
+							HashTable_Add(mapDir, _strdup(curName), inode);
+						}
+						else
+						{
+							inode->st_mode = S_IFREG | 0644;
+						}
+
+						free(curName);
+
+						inode->st_size = (((UINT64)descriptors[lindex].nFileSizeHigh) << 32) |
+						                 ((UINT64)descriptors[lindex].nFileSizeLow);
+						ticks =
+						    (((UINT64)descriptors[lindex].ftLastWriteTime.dwHighDateTime << 32) |
+						     ((UINT64)descriptors[lindex].ftLastWriteTime.dwLowDateTime)) -
+						    116444736000000000;
+						inode->st_mtim.tv_sec = ticks / 10000000ULL;
+						/* tv_nsec Not used for now */
+						inode->st_mtim.tv_nsec = ticks % 10000000ULL;
```

Comment:
I believe these magic constants need to be named and commented on.

Also, I'm not sure that those values are always set in FILEDESCRIPTOR structure.

Reply:
> Also, I'm not sure that those values are always set in FILEDESCRIPTOR structure.

They are not. For example, if there is a server implementation that uses FreeRDPs current implementation for "current-peer to other-peer" file transfer, then they are not set (fixed by https://github.com/FreeRDP/FreeRDP/pull/6704).
The last write time should only be used if the `FD_WRITESTIME` flag is set, otherwise just fall back to `time (NULL)`.